### PR TITLE
dnsdist: Dynamic discovery and upgrade of backends

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Implement https://datatracker.ietf.org/doc/html/draft-ietf-add-ddr-04 to automatically discover if we can upgrade Do53 backends to DoT or DoH.

<strike>Needs regression tests.</strike>

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

**Comment by omoerbeek** [comment]
I'm seeing compile warnings on OpenBSD (clang-13)
```
In file included from dnsdist-idstate.cc:2:
./dnsdist.hh:696:3: warning: explicitly defaulted move constructor is implicitly deleted [-Wdefaulted-function-deleted]
  DownstreamState(DownstreamState&&) = default;
  ^
./dnsdist.hh:766:10: note: move constructor of 'DownstreamState' is implicitly deleted because field 'sendErrors' has a deleted move constructor
  stat_t sendErrors{0};
         ^
./stat_t.hh:43:5: note: 'stat_t_trait' has been explicitly marked deleted here
    stat_t_trait(const stat_t_trait&) = delete;
    ^
In file included from dnsdist-idstate.cc:2:
./dnsdist.hh:698:20: warning: explicitly defaulted move assignment operator is implicitly deleted [-Wdefaulted-function-deleted]
  DownstreamState& operator=(DownstreamState&&) = default;
                   ^
./dnsdist.hh:795:48: note: move assignment operator of 'DownstreamState' is implicitly deleted because field 'hashes' has a deleted move assignment operator
  SharedLockGuarded<std::vector<unsigned int>> hashes;
                                               ^
./lock.hh:450:21: note: copy assignment operator of 'SharedLockGuarded<std::vector<unsigned int>>' is implicitly deleted because field 'd_mutex' has a deleted copy assignment operator
  std::shared_mutex d_mutex;
                    ^
/usr/include/c++/v1/shared_mutex:188:19: note: 'operator=' has been explicitly marked deleted here
    shared_mutex& operator=(const shared_mutex&) = delete;
```

**Comment by rgacogne** [comment]
> I'm seeing compile warnings on OpenBSD (clang-13)
> 
> ```
> In file included from dnsdist-idstate.cc:2:
> ./dnsdist.hh:696:3: warning: explicitly defaulted move constructor is implicitly deleted [-Wdefaulted-function-deleted]
>   DownstreamState(DownstreamState&&) = default;
>   ^
> ./dnsdist.hh:766:10: note: move constructor of 'DownstreamState' is implicitly deleted because field 'sendErrors' has a deleted move constructor
>   stat_t sendErrors{0};
>          ^
> ./stat_t.hh:43:5: note: 'stat_t_trait' has been explicitly marked deleted here
>     stat_t_trait(const stat_t_trait&) = delete;
>     ^
> In file included from dnsdist-idstate.cc:2:
> ./dnsdist.hh:698:20: warning: explicitly defaulted move assignment operator is implicitly deleted [-Wdefaulted-function-deleted]
>   DownstreamState& operator=(DownstreamState&&) = default;
>                    ^
> ./dnsdist.hh:795:48: note: move assignment operator of 'DownstreamState' is implicitly deleted because field 'hashes' has a deleted move assignment operator
>   SharedLockGuarded<std::vector<unsigned int>> hashes;
>                                                ^
> ./lock.hh:450:21: note: copy assignment operator of 'SharedLockGuarded<std::vector<unsigned int>>' is implicitly deleted because field 'd_mutex' has a deleted copy assignment operator
>   std::shared_mutex d_mutex;
>                     ^
> /usr/include/c++/v1/shared_mutex:188:19: note: 'operator=' has been explicitly marked deleted here
>     shared_mutex& operator=(const shared_mutex&) = delete;
> ```

We do not actually use these so I deleted them.

**Comment by rgacogne** [comment]
Rebased to fix several conflicts with #11163.

**Review by Habbie — APPROVED** [review]
Awesome work! Tests and docs look good. Code also looks good (but I did not read every line - a lot appeared to be "mechanical" changes like the new config object). I tested DoT auto-upgrade with quad9 and that works perfectly.

File: pdns/dnsdistdist/docs/reference/config.rst [hunk 1]
```diff
@@ -584,12 +587,18 @@ Servers
...
       dohPath=STRING,           -- Enable DNS over HTTPS communication for this backend, using POST queries to the HTTP host supplied as ``subjectName`` and the HTTP path supplied in this parameter.
       addXForwardedHeaders=BOOL,-- Whether to add X-Forwarded-For, X-Forwarded-Port and X-Forwarded-Proto headers to a DNS over HTTPS backend.
       releaseBuffers=BOOL,      -- Whether OpenSSL should release its I/O buffers when a connection goes idle, saving roughly 35 kB of memory per connection. Default to true.
-      enableRenegotiation=BOOL  -- Whether secure TLS renegotiation should be enabled. Disabled by default since it increases the attack surface and is seldom used for DNS.
+      enableRenegotiation=BOOL, -- Whether secure TLS renegotiation should be enabled. Disabled by default since it increases the attack surface and is seldom used for DNS.
+      autoUpgrade=BOOL,         -- Whether to use the 'Discovery of Designated Resolvers' mechanism to automatically upgrade an Do53 backend to DoT or DoH. Default to false.
```
Comment (RobinGeuze) [inline pdns/dnsdistdist/docs/reference/config.rst:597]: If both are available, which does it use?

File: pdns/dnsdist-console.cc [hunk 2]
```diff
@@ -815,21 +815,27 @@ static void controlClientThread(ConsoleConnection&& conn)
   try
   {
     setThreadName("dnsdist/conscli");
+    vinfolog("Started a new thread to handle the control connection");
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-console.cc:820]: No!! Nice catch :+1:

File: pdns/dnsdistdist/docs/reference/config.rst (hunk 1 above)
Comment (rgacogne) [reply pdns/dnsdistdist/docs/reference/config.rst:597]: I'll add a comment explaining that we respect the priorities from the SVCB record.

**Review by omoerbeek — APPROVED** [review]

**Pull request merged**